File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.cpp
//merge sort o(m+n) tc and o(n) sc

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        vector<int> nums3; 
        
        int n1 = nums1.size();
        int n2 = nums2.size();
        
        int i = 0, j = 0;
        
        while (i < n1 && j < n2) {
            if (nums1[i] <= nums2[j]) {
                nums3.push_back(nums1[i++]);
            } else {
                nums3.push_back(nums2[j++]);
            }
        }
        
        while (i < n1) {
            nums3.push_back(nums1[i++]);
        }
        
        while (j < n2) {
            nums3.push_back(nums2[j++]);
        }
        
        int n = n1 + n2;
        
        if (n % 2 == 1) {
            return static_cast<double>(nums3[n / 2]); // Median for odd number of elements
        } else {
            // Median for even number of elements is the average of the two middle elements
            int middle1 = nums3[n / 2 - 1];
            int middle2 = nums3[n / 2];
            return static_cast<double>(middle1 + middle2) / 2.0;
        }
    }
};

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int n1 = nums1.length, n2 = nums2.length;
        int i = 0, j = 0;
        int n3 = (n1 + n2);

        int ind2 = (n3 / 2);
        int ind1 = ind2 - 1;

        int cnt = 0;

        int EL1 = +1, EL2 = -1;

        while (i < n1 && j < n2) {
            if (nums1[i] < nums2[j]) {
                if (cnt == ind1) {
                    EL1 = nums1[i];
                }
                if (cnt == ind2) {
                    EL2 = nums1[i];
                }
                cnt++;
                i++;
            } else {
                if (cnt == ind1) {
                    EL1 = nums2[j];
                }
                if (cnt == ind2) {
                    EL2 = nums2[j];
                }
                cnt++;
                j++;
            }
        }

        while (i < n1) {
            if (cnt == ind1) {
                EL1 = nums1[i];
            }
            if (cnt == ind2) {
                EL2 = nums1[i];
            }
            cnt++;
            i++;
        }

        while (j < n2) {
            if (cnt == ind1) {
                EL1 = nums2[j];
            }
            if (cnt == ind2) {
                EL2 = nums2[j];
            }
            cnt++;
            j++;
        }

        if (n3 % 2 == 1) {
            return EL2;
        } else {
            return (double) ((double) (EL1 + EL2)) / 2.0;
        }
    }
}
